p2panda-rs/src/secret_group/mls/group.rs

```diff
@@ -0,0 +1,204 @@
+    ///
+    /// New members can use the information inside the Welcome message to set up their own group
+    /// state and derive a shared secret.
+    pub fn new_from_welcome(
+        provider: &impl OpenMlsCryptoProvider,
+        welcome: Welcome,
+    ) -> Result<Self, MlsError> {
+        let group = ManagedGroup::new_from_welcome(provider, &Self::config(), welcome, None)?;
+        Ok(Self(group))
+    }
+
+    // Membership
+    // ==========
+
+    /// Adds new members to an existing MLS group which results in a Commit and Welcome message.
+    ///
+    /// The sender of a Commit message is responsible for sending a Welcome message to any new
+    /// members. The Welcome message provides the new members with the current state of the group,
+    /// after the application of the Commit message.
+    pub fn add_members(
+        &mut self,
+        provider: &impl OpenMlsCryptoProvider,
+        members: &[KeyPackage],
+    ) -> Result<(MlsMessageOut, Welcome), MlsError> {
+        Ok(self.0.add_members(provider, members)?)
+    }
+
+    /// Removes members from a MLS group which results in a Commit message.
+    ///
+    /// Please note that this current implementation requires the member leaf indexes to identify
+    /// the to-be-removed members which will be changed in the future.
+    ///
+    /// See: https://github.com/openmls/openmls/issues/541
+    pub fn remove_members(
+        &mut self,
+        provider: &impl OpenMlsCryptoProvider,
+        member_leaf_indexes: &[usize],
+    ) -> Result<MlsMessageOut, MlsError> {
+        let results = self.0.remove_members(provider, member_leaf_indexes)?;
+
+        // MLS returns an `MlsMessageOut` and optional `Welcome` message when removing a member. We
+        // can be sure there will be no Welcome message in the p2panda case, so we only take the
+        // "out" message.
+        Ok(results.0)
+    }
+
+    // Commits
+    // =======
+
+    /// Processes a Commit message.
+    ///
+    /// A Commit message initiates a new epoch for the group, based on a collection of Proposals.
+    /// It instructs group members to update their representation of the state of the group by
+    /// applying the proposals and advancing the key schedule.
+    pub fn process_commit(
+        &mut self,
+        provider: &impl OpenMlsCryptoProvider,
+        message: MlsMessageIn,
+    ) -> Result<Vec<GroupEvent>, MlsError> {
+        Ok(self.0.process_message(message, provider)?)
+    }
+
+    // Encryption
+    // ==========
+
+    /// Exports secrets based on the current MLS group epoch.
+    ///
+    /// The main MLS key schedule provides an exporter which can be used by an application as the
+    /// basis to derive new secrets outside the MLS layer.
+    pub fn export_secret(
+        &self,
+        provider: &impl OpenMlsCryptoProvider,
+        label: &str,
+        key_length: usize,
+    ) -> Result<Vec<u8>, MlsError> {
+        Ok(self.0.export_secret(provider, label, &[], key_length)?)
```

Comment:
Note that you can add a context to bind the secret to something if you so wish.

Reply:
Is this some sort of AAD? Would it make sense to bind it to a group id?

Reply:
No, an AAD is something that exists in the context of AEAD. This is an additional input parameter that ends up in HKDF as well. It can be used for any context the clients have and it just means they'd implicitly agree on it. Using something group-related like the group ID is unnecessary because the group state is already part of the key schedule. So in general this could be something application-specific.